llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp

```diff
@@ -293,6 +295,36 @@ static bool expandClampIntrinsic(CallInst *Orig, Intrinsic::ID ClampIntrinsic) {
   return true;
 }
 
+static bool expandSignIntrinsic(CallInst *Orig) {
+  IRBuilder<> Builder(Orig->getParent());
+  Value *X = Orig->getOperand(0);
+  Type *Ty = X->getType();
+  Type *ScalarTy = Ty->getScalarType();
+  Type *RetTy = Orig->getType();
+  Constant *Zero = Constant::getNullValue(Ty);
+  Builder.SetInsertPoint(Orig);
```

Comment:
We can either add `usign` builtins/intrinsics all the way down from hlsl to dxil and spirv. Or just add a high level builtin to hlsl. 

I don't like the idea of having an intrinsic for `select(x == 0, 0, 1)`.
I wonder from a consistency point of view if it is even necessary to have a `usign` builtin since `abs` also only seems to offer the signed variants in `hlsl_intrinsics.h`.

Reply:
part 1 fixed in 61bc2a8

Reply:
i'm not opposed to doing a high level operation for unsigned  `sign` cases, however  for reference this is what DXC does: https://github.com/microsoft/DirectXShaderCompiler/blob/0e7591a6ee94c8c8eb0d536ce7815fd56a776451/lib/HLSL/HLOperationLower.cpp#L2703C1-L2714C2

Reply:
Let's do the high level op then.

Reply:
> A small nitpick I like to keep the the builder setup code close together ie
> 
> ```c++
>  IRBuilder<> Builder(Orig->getParent());
> Builder.SetInsertPoint(Orig);
> ```
I didn't see any discussion about this nor any changes in code. Also peculiar to me is the resistance to use the more direct approach which is to set the insertion point with the constructor. Instead of setting it to the start of the basic block and then resetting it, you can just set it with `Orig` and be done with it. 

```c++
  IRBuilder<> Builder(Orig);
```